qiita_pet/test/rest/test_study_preparation.py

```diff
@@ -0,0 +1,180 @@
+                                                      ['bar.txt', 1]],
+                'artifact_name': 'a name is a name'}
+
+        response = self.post(uri, data=body, headers=self.headers, asjson=True)
+        exp = {'message': 'Preparation not found'}
+        self.assertEqual(response.code, 404)
+        obs = json_decode(response.body)
+        self.assertEqual(obs, exp)
+
+    def test_post_unknown_artifact_type(self):
+        uri = '/api/v1/study/1/preparation/1/artifact'
+        body = {'artifact_type': 'foo', 'filepaths': [['foo.txt', 1],
+                                                      ['bar.txt', 1]],
+                'artifact_name': 'a name is a name'}
+
+        response = self.post(uri, data=body, headers=self.headers, asjson=True)
+        self.assertEqual(response.code, 406)
+        obs = json_decode(response.body)
+        self.assertEqual(list(obs.keys()), ['message'])
+        self.assertGreater(len(obs['message']), 0)
+
+    def test_post_unknown_filepath_type_id(self):
+        uri = '/api/v1/study/1/preparation/1/artifact'
+        body = {'artifact_type': 'foo', 'filepaths': [['foo.txt', 123123],
+                                                      ['bar.txt', 1]],
+                'artifact_name': 'a name is a name'}
+
+        response = self.post(uri, data=body, headers=self.headers, asjson=True)
+        self.assertEqual(response.code, 406)
+        obs = json_decode(response.body)
+        self.assertEqual(list(obs.keys()), ['message'])
+        self.assertGreater(len(obs['message']), 0)
+
+    def test_post_files_notfound(self):
+        uri = '/api/v1/study/1/preparation/1/artifact'
+        body = {'artifact_type': 'foo', 'filepaths': [['foo.txt', 1],
+                                                      ['bar.txt', 1]],
+                'artifact_name': 'a name is a name'}
+
+        response = self.post(uri, data=body, headers=self.headers, asjson=True)
+        self.assertEqual(response.code, 406)
+        obs = json_decode(response.body)
+        self.assertEqual(list(obs.keys()), ['message'])
+        self.assertGreater(len(obs['message']), 0)
+
+    def test_post_valid(self):
+        dontcare, uploads_dir = get_mountpoint('uploads')[0]
+        foo_fp = os.path.join(uploads_dir, '1', 'foo.txt')
+        bar_fp = os.path.join(uploads_dir, '1', 'bar.txt')
+        with open(foo_fp, 'w') as fp:
+            fp.write("@x\nATGC\n+\nHHHH\n")
+        with open(bar_fp, 'w') as fp:
+            fp.write("@x\nATGC\n+\nHHHH\n")
+
+        prep = StringIO(EXP_PREP_TEMPLATE.format(1))
+        prep_table = load_template_to_dataframe(prep)
+
+        response = self.post('/api/v1/study/1/preparation?data_type=16S',
+                             data=prep_table.T.to_dict(),
+                             headers=self.headers, asjson=True)
+        prepid = json_decode(response.body)['id']
+
+        uri = '/api/v1/study/1/preparation/%d/artifact' % prepid
+        # 1 -> fwd or rev sequences in fastq
```

Comment:
I don't think passing the ids is required. IIRC you can pass the actual string and the code internally will change to the ID. So instead of 3 you can pass 'raw_barcodes'